vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp

```diff
@@ -0,0 +1,312 @@
+}
+
+void
+DataStoreFixedSizeHashTest::insert(uint32_t key)
+{
+    auto hash_map = _hash_map.load(std::memory_order_relaxed);
+    if (hash_map->full()) {
+        grow();
+        hash_map = _hash_map.load(std::memory_order_relaxed);
+    }
+    MyCompare comp(_store, key);
+    std::function<EntryRef(void)> insert_entry([this, key]() -> EntryRef { return _allocator.allocate(key); });
+    auto& result = hash_map->add(comp, insert_entry);
+    auto ref = result.first.load_relaxed();
+    auto &wrapped_entry = _allocator.get_wrapped(ref);
+    EXPECT_EQ(key, wrapped_entry.value());
+}
+
+void
+DataStoreFixedSizeHashTest::remove(uint32_t key)
+{
+    MyCompare comp(_store, key);
+    auto hash_map = _hash_map.load(std::memory_order_relaxed);
+    auto result = hash_map->remove(comp, EntryRef());
+    if (result != nullptr) {
+        auto ref = result->first.load_relaxed();
+        auto &wrapped_entry = _allocator.get_wrapped(ref);
+        EXPECT_EQ(key, wrapped_entry.value());
+        _allocator.hold(ref);
+    }
+}
+
+void
+DataStoreFixedSizeHashTest::read_work(uint32_t cnt)
+{
+    vespalib::Rand48 rnd;
+    long found = 0;
+    rnd.srand48(++_read_seed);
+    uint32_t i;
+    for (i = 0; i < cnt && _stop_read.load() == 0; ++i) {
+        auto guard = _generation_handler.takeGuard();
+        uint32_t key = rnd.lrand48() % (_keyLimit + 1);
+        MyCompare comp(_store, key);
+        auto hash_map = _hash_map.load(std::memory_order_acquire);
+        auto result = hash_map->find(comp, EntryRef());
+        if (result != nullptr) {
+            auto ref = result->first.load_relaxed();
+            auto &wrapped_entry = _allocator.get_wrapped(ref);
+            EXPECT_EQ(key, wrapped_entry.value());
+            ++found;
+        }
+    }
+    _done_read_work += i;
+    _found_count += found;
+    LOG(info, "done %u read work", i);
+}
+
+
+void
+DataStoreFixedSizeHashTest::read_work()
+{
+    read_work(std::numeric_limits<uint32_t>::max());
+}
+
+
+void
+DataStoreFixedSizeHashTest::write_work(uint32_t cnt)
+{
+    vespalib::Rand48 &rnd(_rnd);
+    for (uint32_t i = 0; i < cnt; ++i) {
+        uint32_t key = rnd.lrand48() % _keyLimit;
+        if ((rnd.lrand48() & 1) == 0) {
+            insert(key);
+        } else {
+            remove(key);
+        }
+        commit();
+    }
+    _done_write_work += cnt;
+    _stop_read = 1;
+    LOG(info, "done %u write work", cnt);
+}
+
+
+TEST_F(DataStoreFixedSizeHashTest, smoke_test)
+{
+    EXPECT_EQ(0, size());
+    insert(1);
+    EXPECT_EQ(1, size());
+    remove(2);
+    EXPECT_EQ(1, size());
+    insert(1);
+    EXPECT_EQ(1, size());
+    insert(5);
+    EXPECT_EQ(2, size());
+    insert(4);
+    EXPECT_EQ(3, size());
+    remove(3);
+    EXPECT_EQ(3, size());
+    remove(5);
+    EXPECT_EQ(2, size());
+    commit();
+    MyCompare comp3(_store, 3);
+    auto hash_map = _hash_map.load(std::memory_order_acquire);
+    auto result3 = hash_map->find(comp3, EntryRef());
```

Comment:
Consider creating helper function(s) to verify find().

Reply:
Done.